### PR TITLE
Extended config properties

### PR DESCRIPTION
**Pull request by jack-berg**
Replaces draft #5873.

@open-telemetry/java-instrumentation-maintainers in the 9/21 Java SIG meeting I brought up bringing the file configuration prototype into the java agent, and received feedback that it will be important to specify other java agent properties besides the SDK in a configuration. Examples include accessing [jmx](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jmx-metrics/javaagent) configuration, and other standard properties like `otel.javaagent.debug=true`. 

This PR extends file configuration tooling to include access to configuration not part of the [configuration schema](https://github.com/open-telemetry/opentelemetry-configuration). Summary:
- A user specifies both SDK and java agent configuration in a single YAML file, and sets `OTEL_CONFIG_FILE` to the file path
- Autoconfigure detects that `OTEL_CONFIG_FILE` is set and that `opentelemetry-sdk-extension-incubator` is on the classpath, and:
  - Parses / interprets the config file to configure the SDK
  - Converts the `OpenTelemetryConfiguration` model to a generic `ExtendedConfigProperties` representation accessible via `AutoConfiguredOpenTelemetrySdk#getConfig()`
- `ExtendedConfigProperties` implements `ConfigProperties` and adds some additional methods for accessing nested complex configuration types (nested maps, and lists of maps)
- The opentelemetry java agent accesses `ExtendedConfigProperties` and uses it for configuration

Suppose you have a configuration file with contents like:
```
file_format: "0.1"
tracer_provder:
  processors:
    - batch:
        exporters:
          otlp:
javaagent:
  debug: true
  jmx:
    foo: bar
```

In java, you use the file to configure the SDK, as well as access the extended "javaagent" portion of the schema as follows:
```
export OTEL_CONFIG_FILE=/path-to-config.yaml
...
AutoConfiguredOpenTelemetrySdk autoConfiguredSdk = AutoConfiguredOpenTelemetry.builder().build();
ExtendedConfigProperties configProps = (ExtendedConfigProperties) autoConfiguredSdk.getConfig();

ExtendedConfigProperties javaagentProps = configProps.getConfigProperties("javaagent");
Boolean debug = javaagentProps.getBoolean("debug");
ExtendedConfigProperties jmxProps = javaagentProps.getConfigProperties("jmx");
String foo = jmxProps.getString("foo");

assert foo.equals("bar");

// Or use dot notation 
assert configProps.get("javaagent.jmx.foo").equals("bar");
assert configProps.get("javaagent.debug").equals(true);
```

### CODE REVIEW

**Comment by jack-berg** [comment]
PS. [opentelemetry-java-examples#227](https://github.com/open-telemetry/opentelemetry-java-examples/pull/227) how this code can be coupled with changes to the contrib [RuleBasedRoutingSampler](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/samplers/src/main/java/io/opentelemetry/contrib/sampler/RuleBasedRoutingSampler.java) to be able to configure the otel java agent to drop spans matching a particular pattern (like health checks), which is a very popular feature request.

File: sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigPropertiesTest.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+          + "      str_key2: str_value2\n"
+          + "      int_key2: 3\n"
+          + "  list_key:\n"
+          + "    - str_key1: str_value1\n"
+          + "      int_key1: 2\n"
+          + "      map_key1:\n"
+          + "        str_key2: str_value2\n"
+          + "        int_key2: 3\n"
+          + "    - str_key1: str_value1\n"
+          + "      int_key1: 2";
```
Comment (jack-berg) [inline sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigPropertiesTest.java:53]: Check out this test for examples of how to use the API to read complex types defined in a configuration file.

Notably, I've added support for dot notation to allow existing system property / environment variable configuration to be translated to a hierarchical representation in YAML.

For example, the property `otel.javaagent.debug` would be represented in YAML with:
```
otel:
  javaagent:
    debug: true
```
And you could access it using `configProperties.getBoolean("otel.javaagent.debug")`.

File: sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/FileConfigurationTest.java [hunk 2]
```diff
@@ -175,4 +180,27 @@ void configFile_Error(@TempDir Path tempDir) throws IOException {
         .isInstanceOf(ConfigurationException.class)
         .hasMessage("Unrecognized span exporter(s): [foo]");
   }
+
+  @Test
+  void configFile_ExtendedConfigProperties() {
+    ConfigProperties config =
+        DefaultConfigProperties.createFromMap(
+            Collections.singletonMap("OTEL_CONFIG_FILE", configFilePath.toString()));
```
Comment (brunobat) [inline sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/FileConfigurationTest.java:188]: We need to provide a way to deactivate `OTEL_CONFIG_FILE` config loading if the properties supplier is being used directly in the Autoconfiguration.

**Review by brunobat — CHANGES_REQUESTED** [review]
We need to clarify:
* Which properties will work in the file configuration but not as env. vars.
* We need to make sure all properties can be set through the properties supplier and that the file configuration can be ignored in that case.

**Comment by jack-berg** [comment]
> We need to clarify:

> Which properties will work in the file configuration but not as env. vars.
> We need to make sure all properties can be set through the properties supplier and that the file configuration can be ignored in that case.

@brunobat That feedback is orthogonal to this PR. This PR is about allowing complex configuration to be accessible via an `ExtendedConfigProperties` such that SPI implementations can reach more than just simple types.

File: sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/FileConfigurationTest.java (hunk 2 above)
Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/FileConfigurationTest.java:188]: This is unrelated to this PR. Its already possible to deactivate `OTEL_CONFIG_FILE` via property suppliers. See the [source code](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java#L387) for how `OTEL_CONFIG_FILE` is loaded. Its accessed via `ConfigProperties.getString("otel.config.file")` using a `ConfigProperties` instance that is resolved using all the property suppliers and customizers typical in autoconfigure.

**Pull request closed without merging**